### DISCUSSION TITLE
Can we rely on the private `@_parent`?

### DISCUSSION THREAD

**Discussion by ianterrell** [yippee-fun/phlex#510] [Component architecture]
I've been playing with Phlex and enjoying it! It's really great to see innovation in this space.

One thing I'm trying to do is theme components. With a lot of CSS that is possible fairly easily with wrapping classes, but I'm using Tailwind with a custom version of something like [`class_variants`](https://github.com/avo-hq/class_variants) to pick which classes to use amongst various styles. One component, many configurable styles — sometimes grouped in an overarching theme.

I wondered if I could inspect the view hierarchy, and digging through the source I noticed a helpful looking `@_parent` in the private part of the API...

https://github.com/joeldrapper/phlex/blob/0481870c9267139dec91a3245d2dc95084e71700/lib/phlex/sgml.rb#L47

With that I can set up something that will look for a theme along the rendered view hierarchy via ActiveSupport's [`class_attribute`](https://api.rubyonrails.org/classes/Class.html#method-i-class_attribute):

```ruby
class ApplicationComponent < Phlex::HTML
  # ...

  class_attribute :theme, instance_accessor: false

  def theme
    self.class.theme || @_parent&.theme
  end

  def classes(key, theme:)
    # look up classes
  end

  def text_field(form, attribute, **options)
    render Form::TextFieldComponent.new(form, attribute, **options)
  end
end

class Form::TextFieldComponent < ApplicationComponent
  # ...

  def template
    @form.text_field @attribute, classes(:text_field, theme:)
  end
end
```

And then use it something like this:

```ruby
class Articles::FormView < ApplicationView
  self.theme = :user

  # ...

  def template
    # ...

    form_with model: @article do |form|
      text_field form, :title
      text_field form, :body, description: true
    end
  end
end
```

Now views can specify a theme and all the components get rendered with the appropriate styles since they can traverse up the hierarchy.

It may be a terrible idea to be doing this, but it seems to work and completely change the look of a page with just a line! I like that part of it.

Does anyone have any thoughts? Is the view render tree parentage something that could be moved into a supported public API?

**Comment by joeldrapper** [comment]
Hey @ianterrell, the `@_parent` instance variable was required for a feature at one point. I don't believe it's required anymore, but I didn't remove it as I’m considering making it part of the public API.

I think this use-case makes a lot of sense. I would definitely consider caching the theme at each layer with `||=`.

```ruby
def theme
  if defined?(@theme)
    @theme
  else
    @theme = @_parent&.theme
  end
end
```

I would also skip the class attribute entirely. You can specify a theme with a single-line method definition like this instead.

```ruby
def theme = :foo
```

If `@_parent` becomes part of the public API, it will probably be through a `parent` method. Anything with an underscore is private and could change in a patch update.

**Comment by ianterrell** [comment]
> I would also skip the class attribute entirely. 

For sure! That was silly of me. :)

> If @_parent becomes part of the public API, it will probably be through a parent method. Anything with an underscore is private and could change in a patch update.

Definitely on the `_` — just bringing up the general topic. Would love to see a supported method for this feature. 

I didn't dig into it, but it seemed that the layout was outside of this rendering tree; I was hoping to put a theme in the layout at first.

Danke!

**Reply by joeldrapper** [reply]
There are a couple of other ways to do layouts besides the Rails layout system.

1. You could define a layout as a module that defines `around_template` and have your views include that module.
2. You could have your views render a layout component and pass everything into that component's content block.
